header/p2p/peer_tracker.go

```diff
@@ -63,42 +100,91 @@ func (p *peerTracker) connected(pID peer.ID) {
 	if p.host.ID() == pID {
 		return
 	}
+	p.Lock()
+	defer p.Unlock()
+	// skip adding the peer to avoid overfilling of the peerTracker with unused peers if:
+	// peerTracker reaches the maxTrackerSize and there are more connected peers
+	// than disconnected peers.
+	if len(p.trackedPeers)+len(p.disconnectedPeers) > p.maxPeerTrackerSize &&
+		len(p.trackedPeers) > len(p.disconnectedPeers) {
+		return
+	}
+
 	for _, c := range p.host.Network().ConnsToPeer(pID) {
 		// check if connection is short-termed and skip this peer
 		if c.Stat().Transient {
 			return
 		}
 	}
```

Comment:
Just realized that here we have a bug, rare but still. There can be two connections between two peers, one of which is transient, this logic aborts fully if one of them is transient.

Reply:
Another suggestion is to check connections first, and only then maxPeerTrackerSize

Reply:
> Just realized that here we have a bug, rare but still. There can be two connections between two peers, one of which is transient, this logic aborts fully if one of them is transient.

Actually I discussed it with you. And we agreed to skip the peer.

Reply:
> Another suggestion is to check connections first, and only then maxPeerTrackerSize

Why? if our peerTracker is full, then we do not need to go further.